File: my_cpp_pkg/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

class AddTwoIntsClientNode : public rclcpp::Node // MODIFY NAME
{
public:
    AddTwoIntsClientNode() : Node("add_two_ints_client") // MODIFY NAME
    {   
        // We call the service in a new thread, that could possibly get stuck, but we 
        // can continue spinning
        threads_.push_back(std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 1, 4)));
        threads_.push_back(std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 5, 7)));
    }

    void callAddTwoIntsService(int a, int b)
    {   
        auto client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
        while(!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for the server to be up...");
        }

        auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = a;
        request->b = b;

        auto future = client->async_send_request(request);
        // Here to wait for the future we use future.get() which waits until the future is completed
        try
        {
            // this will block the thread while waiting
            auto response = future.get();
            RCLCPP_INFO(this->get_logger(), "%d + %d = %d", a,b, response->sum);
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Service call failed.");
        }
    }

private:
    std::vector<std::thread> threads_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsClientNode>(); // MODIFY NAME
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: my_py_pkg/my_py_pkg/add_two_ints_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts
from functools import partial

class AddTwoIntsClientNode(Node): # MODIFY NAME
    def __init__(self):
        super().__init__("add_two_ints_client") # MODIFY NAME
        self.call_add_two_ints_server(6,7)
        self.call_add_two_ints_server(2,3)


    # FROM HERE, WE CAN ALWAYS USE THE SAME CODE FOR THE SERVICES
    # We don't put the service in the __init__ but we create a new function
    # so that we can call it multiple times
    def call_add_two_ints_server(self, a, b):
        client = self.create_client(AddTwoInts, "add_two_ints")
        while not client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for Server Add Two Ints...")
        request = AddTwoInts.Request()
        request.a = a
        request.b = b

        future = client.call_async(request)
        # When the future is complete, callback is called. We have to do like this, 
        # because we can't spin here (we already spin in the main)
        # NB: The callback is called with the Future object as its only argument.
        # Partial(func, specific_param) set the param of the func to specific_parm
        # (it goes from left to right OR I can explicitly assign them)
        future.add_done_callback(partial(self.callback_call_add_two_ints, a=a, b=b))
    
    def callback_call_add_two_ints(self, future, a, b):
        try:
            response = future.result()
            self.get_logger().info(
                str(a)+"+"+str(b)+"="+str(response.sum))
        except Exception as e:
            self.get_logger().error("Service call failed %r"%(e,))

def main(args=None):
    rclpy.init(args=args)
    node = AddTwoIntsClientNode() # MODIFY NAME
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
